### PR TITLE
Zero123 log weighted loss terms and optional gradient accumulation

### PR DESCRIPTION
**Pull request by claforte**
Support gradient accumulation in Zero123
- "accumulate" combines in 1 forward/backward pass the ref and zero123 guidance
  - I think that this should improve Adan momentum (compared with alternating between very different loss terms every batch) and support higher learning rate... but that remains to be proven conclusively
- defaults to "alternate" (i.e. previous default) but I think, after suitable hyperparam tuning, the accumulate approach will outperform "alternate".

Log (e.g. in wandb and CSV) each weighted loss term, to make it much easier to visualize their respective importance, e.g.:
image

From the above figure, my take-aways/next steps are:
- disable ref sparsity loss?
- 3d normal smooth seems set too high ... divide by 3?
  - this explains why Anya seem overly blocky regardless of "accumulate" or "alternate", e.g. for alternate:
- what do opaque and orient terms do?

Further comparison between "accumulate" and "alternate":
- 1.84 iters/s with alternate, 1.05 iters/s with accumulate
- 24GB VRAM with alternate, 29GB VRAM with accumulate

Initial conclusions quality-wise:
 - in wall-time, accumulate seems to converge a tiny bit faster to equivalent visual quality (early on)
 - Later in training, it becomes harder to compare the two fairly... since "alternate" runs through iterations faster, it hits extra levels of hash encoding faster (wall-time), and never gets rid of the extra noise.
 - I should have halved the number of iterations for "accumulate"

Runs:
- https://stability.wandb.io/threestudio/claforte-gradient_accum

### CODE REVIEW

**Comment by claforte** [comment]
@bennyguo I'd like your review... short-term this works for Zero123, but soon after, we'll need to implement similar features to combine SD- and  DeepFloyd-guidance with ref image... I think we can come up with a cleaner, orthogonal approach, e.g. one config file that refers to 2 models and accumulate their gradients or alternate. Might be faster to talk in person, e.g. on my Friday morning / your evening (if you can spare the time)

**Comment by voletiv** [comment]
Highly agreed!

**Review by bennyguo — APPROVED** [review]

File: threestudio/systems/zero123.py [hunk 1]
```diff
@@ -112,48 +109,37 @@ def training_step(self, batch, batch_idx):
...
-                **batch,
-                rgb_as_latents=False,
-                guidance_eval=guidance_eval,
-            )
+                ))
+        elif guidance == "zero123":
+            # zero123
+            guidance_out, guidance_eval_out = self.guidance(out["comp_rgb"], **batch, rgb_as_latents=False, guidance_eval=guidance_eval)
+            # claforte: TODO: rename the loss_terms keys
+            set_loss("sds", guidance_out["loss_sds"])
```
Comment (voletiv) [inline threestudio/systems/zero123.py:119]: Good enough for now, but will need to add another `set_loss` line when using prolific_dreamer

**Comment by claforte** [comment]
Thanks! I'll rebase-and-merge since I'm pretty sure each commit can run on its own.
